### PR TITLE
Fixed #30944 -- Changed reusable apps docs to use a declarative config.

### PR DESCRIPTION
**Pull request by jdufresne**
https://code.djangoproject.com/ticket/30944

### CODE REVIEW

File: docs/intro/reusable-apps.txt [hunk 1]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
    license. Just be aware that your licensing choice will affect who is able
    to use your code.
 
-#. Next we'll create a ``setup.py`` file which provides details about how to
-   build and install the app.  A full explanation of this file is beyond the
-   scope of this tutorial, but the `setuptools docs
-   <https://setuptools.readthedocs.io/en/latest/>`_ have a good
-   explanation. Create a file ``django-polls/setup.py`` with the following
-   contents:
+#. Next we'll create a ``setup.cfg`` and ``setup.py`` file which provides
```
Comment (claudep) [inline docs/intro/reusable-apps.txt]: `create ``setup.cfg`` and ``setup.py`` files which provide`

File: docs/intro/reusable-apps.txt [hunk 2]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
+   details about how to build and install the app. A full explanation of this
+   file is beyond the scope of this tutorial, but the `setuptools docs
```
Comment (claudep) [inline docs/intro/reusable-apps.txt]: `these files`

**Comment by jdufresne** [comment]
Thanks. Fixed up the grammar.

File: docs/intro/reusable-apps.txt [hunk 3]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
-#. Next we'll create a ``setup.py`` file which provides details about how to
-   build and install the app.  A full explanation of this file is beyond the
-   scope of this tutorial, but the `setuptools docs
-   <https://setuptools.readthedocs.io/en/latest/>`_ have a good
-   explanation. Create a file ``django-polls/setup.py`` with the following
-   contents:
+#. Next we'll create ``setup.cfg`` and ``setup.py`` files which provide details
+   about how to build and install the app. A full explanation of these files is
+   beyond the scope of this tutorial, but the `setuptools docs
+   <https://setuptools.readthedocs.io/en/latest/>`_ have a good explanation.
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: `…setuptools docs have…` → `…setuptools documentation has…`

File: docs/intro/reusable-apps.txt [hunk 4]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
+        classifiers =
+            Environment :: Web Environment
+            Framework :: Django
+            Framework :: Django :: X.Y  # replace "X.Y" as appropriate
+            Intended Audience :: Developers
+            License :: OSI Approved :: BSD License  # example license
+            Operating System :: OS Independent
+            Programming Language :: Python
+            Programming Language :: Python :: 3.6
+            Programming Language :: Python :: 3.7
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: Please also add the classifier for Python 3.8 to match supported versions for 3.1.

(Care should be taken to remove it again if these changes are backported.)

File: docs/intro/reusable-apps.txt [hunk 5]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
+
+        [metadata]
+        name = django-polls
+        version = 0.1
+        description = A Django app to conduct Web-based polls.
+        long_description = file: README.rst
+        url = https://www.example.com/
+        author = Your Name
+        author_email = yourname@example.com
+        license = BSD License  # example license
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: ~~This is not required if the trove classifier below is specific enough. Not sure whether we want to mention that...~~

I guess that would beyond the scope as mentioned in the preceding paragraph, and also there is no specific enough classifier for BSD licenses.

File: docs/intro/reusable-apps.txt [hunk 6]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
+            Operating System :: OS Independent
+            Programming Language :: Python
+            Programming Language :: Python :: 3.6
+            Programming Language :: Python :: 3.7
+            Topic :: Internet :: WWW/HTTP
+            Topic :: Internet :: WWW/HTTP :: Dynamic Content
+
+        [options]
+        include_package_data = True
+        packages = :find
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: Is this meant to be `find:`?

File: docs/intro/reusable-apps.txt [hunk 7]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
-       :caption: django-polls/setup.py
-
-       import os
-       from setuptools import find_packages, setup
-
-       with open(os.path.join(os.path.dirname(__file__), 'README.rst')) as readme:
-           README = readme.read()
-
-       # allow setup.py to be run from any path
-       os.chdir(os.path.normpath(os.path.join(os.path.abspath(__file__), os.pardir)))
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: This is a good example of weird hackery that we want to get rid of!

File: docs/intro/reusable-apps.txt (hunk 4 above)
Comment (pope1ni) [reply docs/intro/reusable-apps.txt]: We should probably also add `Programming Language :: Python :: 3 :: Only`.

File: docs/intro/reusable-apps.txt [hunk 8]
```diff
@@ -181,50 +181,48 @@ this. For a small app like polls, this process isn't too difficult.
...
+            License :: OSI Approved :: BSD License  # example license
+            Operating System :: OS Independent
+            Programming Language :: Python
+            Programming Language :: Python :: 3.6
+            Programming Language :: Python :: 3.7
+            Topic :: Internet :: WWW/HTTP
+            Topic :: Internet :: WWW/HTTP :: Dynamic Content
+
+        [options]
+        include_package_data = True
```
Comment (pope1ni) [inline docs/intro/reusable-apps.txt]: `True` → `true`.

**Comment by jdufresne** [comment]
Thanks for the review! I've made all suggested changes.

File: docs/intro/reusable-apps.txt (hunk 5 above)
Comment (pope1ni) [reply docs/intro/reusable-apps.txt]: Just to be a bore, based on my https://github.com/django/django/pull/12013#discussion_r342001037, let's make this `BSD-3-Clause` to set a good example. :angel:

Comment (jdufresne) [reply docs/intro/reusable-apps.txt]: That works for me. Done in the latest revision.

**Comment by pope1ni** [comment]
Hi @carltongibson 

So from my perspective PEP 518 is already working effectively from the perspective of pip installing dependencies for build. It looks as though [`setuptools`](https://github.com/pypa/setuptools/blob/5af71d2c5c8015d5520b9aae4a6117d31cefeeac/pyproject.toml#L1-L2) and [`pip`](https://github.com/pypa/pip/blob/7d5967906e660ebecc1b1642f580ee7e25803ccf/pyproject.toml#L1-L2) are swallowing their own medicine.

Some projects such as [`black`](https://github.com/psf/black/blob/9b484d1bcc2e15dcd5544cddab729c76b4d1d2e9/pyproject.toml#L36-L38) are already opting into PEP 517 - that is, adding `build-backend`. I'm not suggesting we go that far yet. I've been using it myself and it mostly works fine, even with editable builds, but there are still some issues that need to be ironed out. The most frustrating thing is that I find it hard to know what the plan is w.r.t. python packaging. There is no clear roadmap, or as least it is not communicated well enough outside of the PyPA and changes keep happening left, right and centre - often breaking things. I wish they'd just have a single fount of all knowledge with a state of the nation and what issues are being addressed and when.

> Should we not be pointing folks to that, or is that currently ahead of the curve?

I think that it'll stay and we're safe enough telling them to add a `pyproject.toml` too.

> I (finally) notice the `provisional` flag on the PEP... I guess that answers my question.

Yeah. I find that weird. It's "provisional", but everyone has started using it. To abandon now would just make a mess and fragment the state of packaging even more.

Cheers.

**Comment by carltongibson** [comment]
> Yeah. I find that weird. It's "provisional", but everyone has started using it. To abandon now would just make a mess and fragment the state of packaging even more.

Yes. 😬 (But isn't this exactly what's happened with `pipenv`? It's blessed for a few months, everyone jumps on, with PRs to every package, calling _one_ names for **not** adopting it instantly, then all of a sudden, flaws are realised and we now how N+1 not-quite-right options... — that's a caricature obviously, but...)

I'm happy for us _qua_ Django to be a little bit behind the fashion. (_Qua_ my own package it's bleeding edge or go home, but that's different... 🙂)

**Comment by pope1ni** [comment]
> Yes.  (But isn't this exactly what's happened with `pipenv`? It's blessed for a few months, everyone jumps on, with PRs to every package, calling _one_ names for **not** adopting it instantly, then, all of a sudden, flaws are realised, and we now how N+1 not-quite-right options... — that's a caricature obviously, but...)

I never used `pipenv`. To me it was just another project that handled dependency management it's own way without any defined standard. PEP 518 (and PEP 517 which I'm not proposing we opt into yet due to issues with editable installs) is a defined standard actually implemented in `pip`, the de facto tool for installing packages.

> I'm happy for us _qua_ Django to be a little bit behind the fashion.

I understand your reservations. I will point out, however, that Django 3.1 is still another 9 months away, so PEP 518 support added in `pip` v10.0.0 will be ~2.5 years old by then. That's surely not jumping the gun?

> (_Qua_ my own package it's bleeding edge or go home, but that's different... )

Heh. Me too!

**Pull request merged**